delayed-state-rendering.md

```diff
@@ -0,0 +1,273 @@
+
+1. what section of an sls file to delay the render of, and
+2. when to start the render and execution of these delayed sections.
+
+To do this, we add a new requisite called `delayed_render`, that takes a list of things to render and execute after the state that has the requisite in it finishes. This list contains either 1) blocks enclosed by a custom jinja tag, similar to `block`, called `delayed_block`, or 2) the names of entire sls files, in the same way as given in other requisites.
+
+It is important to note that we don't want to mark a state directly to indicate its render should be delayed. Instead, we want the ability to mark *Jinja* as needing to be rendered later. We don't want to render a single state later - we want to render Jinja later that contains a state, or states. Doing this allows for loops, conditionals, [Salt's special context vars like `salt` and `grains`](https://docs.saltstack.com/en/latest/ref/states/vars.html), and the full power of the render machinery. This allows for accessing and building logic around data at delayed render time, where that data did not exist at the render of the calling state.
+
+Here are several examples of how state files could use this
+
+### Simple example calling a jinja block:
+
+```salt
+# init.sls
+my_state1:
+  mod.fun1:
+    - do_something
+    - delayed_render:
+      - block: dependant_block_A
+
+{% delayed_block dependant_block_A %}
+my_state2:
+  mod.fun2:
+    - {{ data }}
+{% end_delayed_block %}
+```
+
+In this example, `my_state` indicates that `dependant_block_A` will be rendered and its states executed after it finishes. In this way, `delayed_render` can be thought of as a new kind of requisite. It is similar to `require_in` in that it helps determine an order of execution, and indicates what comes later. It is different, in that `delayed_render` determines what will be executed **immediately after** this state. It also does not simply point to another state. Instead, `delayed_render` points to a block of Jinja defined by the custom Jinja tag `delayed_block`. This block could contain any template code.
+
+Upon initial render of this state (as in a highstate or otherwise), the `delayed_block` Jinja tag effectively cuts all of the text in the block, and stores it in memory on the Salt master, and associates it with the original job. When `my_state1` finishes, the minion calls back to the master, indicating that the `delayed_block` `dependant_block_A` should be rendered, and any states within it executed, on the same minion.
+
+### Simple example calling an sls file:
+
+```salt
+# init.sls
+my_state1:
+  mod.fun1:
+    - do_something
+    - delayed_render:
+      - sls: my_state_file
+```
+
+```salt
+# my_state_file.sls
+my_state2:
+  mod.fun2:
+    - {{ data }}
+```
+
+This is a very similar example, except that instead of a `delayed_render` indicating a particular `delayed_block`, `delayed_render` specifies an entire sls file to render and execute when `my_state1` finishes.
```

Comment:
Not the clearest
```suggestion
This is a similar example, instead of  a `block` tag indented under `delayed_render` targeting a state in the *same* file; you have a 'sls' tag enabling you to target a *separate* state file.
```

Reply:
Changed locally.